### ISSUE TITLE
[DX] Introduce a new `user_permission_get_info()` function

### ISSUE THREAD

**Issue by klonos** [backdrop/backdrop-issues#4097]
Modules can specify permissions via [`hook_permission()`](https://api.backdropcms.org/api/backdrop/1/search/hook_permission):

```php
function my_awesome_module_permission() {
  return array(
    'permission x' => array(
      'title' => t('Permission to do x'),
      'description' => t('Allows people to do x.'),
      'restrict access' => TRUE,
      'warning' => t('Allows people to do ...'),
    ),
    'permission y' => array(
      'title' => t('Permission to do y'),
      'description' => t('Allows people to do y.'),
    ),
  );
}
```

Now, we have [`user_access()`](https://api.backdropcms.org/api/backdrop/1/search/user_access), which allows us to determine if a user has a specific permission:

```php
$user_can_do_x => user_access('permission x');
$user_can_do_y => user_access('permission y');
```

...and we also have [`user_permission_get_modules()`](https://api.backdropcms.org/api/backdrop/1/search/user_permission_get_modules), which returns all permissions, along with which module implements each permission.

...but we don't have any function that returns the properties of a permission. So each time we need to reference a permission say in some help text, we do this:

```php
$form['whatever'] = array(
  ...
  '#description' => t('This option depends on the "Permission to do y" permission.',
  ...
);
```

This is OK, but if at some point we rename the permission (change it's `'title'`), then we need to do a search and replace (at best), in order to find all instances of the literal permission title in the codebase, and update it to the new name/label. Ideally though, we should be able to do something like this:

```php
$permission_info = user_permission_get_info('permission x');
$form['whatever'] = array(
  ...
  '#description' => t('This option depends on the @permission permission.', array('@permission' => $permission_info['title'],
  ...
);
```

That way, we would only need to update any permission info in a single place (in its `hook_permission()` function), which also makes translators' life easier.

As a developer, at minimum I would like this new function to return:

- the title of the permission
- the description of the permission

...additional nice-to-have's:

- an admin path, which can be used to provide a link to `/admin/config/people/permissions#permission x` (scrolls the permissions listing page to the specific permission)
- the module that specifies the hook of the permission (alternatively, have `user_permission_get_modules()` accept an optional `$permission` parameter, so to be able to do `user_permission_get_modules('permission x')`). This could also be used as a search/filter query when #980 gets implemented 😉 
- a list of all user roles this permission has been assigned to (we have [`user_roles()`](https://api.backdropcms.org/api/backdrop/1/search/user_roles) for that - still nice to have)

**Comment by klonos** [comment]
I can across this again while working in the PR for #5885.

This is what I ended up doing this time:
```php
...
  $roles = user_roles();
  $user_module_permissions = module_invoke('user', 'permission');
  $some_message = ' ' . t('Only roles with the %permission permission can configure this.', array('%permission' => $user_module_permissions['administer permissions']['title']));
...
```

It would have been useful to have something like a `user_permission_get_info('administer permissions', 'title')` function.

I might give this a crack.

**Comment by kiamlaluno** [comment]
It would be helpful every time a message shown to the users needs to reference a permission. Instead of hard-casting the permission name, the code would just use the information returned by `hook_permission()`.

**Comment by klonos** [comment]
Yes, thanks @kiamlaluno ...this is one of the use cases I had in mind when suggesting this and working on a PR: https://github.com/backdrop/backdrop/pull/4275

The PR does the following:
- introduces a new `user_permission_get_info()` function, which **_optionally_** accepts a `$permission_name` and permission `$property` as parameters.
- makes `user_permission_get_modules()` a "wrapper function" around the newly-introduced `user_permission_get_info()` function (it basically calls `return user_permission_get_info(NULL, 'module');`)
- introduces a 3 other helper functions (which are also wrapper functions around `user_permission_get_info()`), which help retrieve common permission properties (permission human-readable title and the machine name of the module providing the permission).

So you can now do the following things:
```php
dpm(user_permission_get_modules(), 'all permission modules');
dpm(user_permission_get_info(), 'all permissions and all their properties');
dpm(user_permission_get_info(NULL, 'title'), 'all permission titles');
dpm(user_permission_get_info(NULL, 'module'), 'all permission modules');
dpm(user_permission_get_info('administer permissions'), 'all properties of a specific permission');
dpm(user_permission_get_info('administer permissions', 'title'), 'a specific property of a specific permission');
dpm(user_permission_get_info('administer permissions', 'what'), 'specific property that does not exist');
dpm(user_permission_get_info('nope'), 'specific permission that does not exist');
```

Any feedback is welcomed, but also specifically about these things:
- Thinking to further improve this by caching the results, but not sure about it.
- Should we keep the "singular" variants of the new wrapper functions for retrieving the module and title of a single permission? I added those mostly for consistency and convenience, as it didn't seem appropriate to be calling a function called `user_permission_get_modules()` (plural "modules") and only be returning information for a single permission, in which case it would only be a single module - NOT a list of modules (as it would happen if retrieving that information for all permissions).

**Comment by kiamlaluno** [comment]
As function names, I would use `user_permissions_get_xyz()`, since the returned information is about one or more permissions.

**Comment by klonos** [comment]
Let the naming bike-shading begin! 😅 

Thanks @kiamlaluno ...I'll wait for feedback from more people before I start renaming things.
